File: backend/app/services/weather_service.py
"""
Weather service for STS Clearance system
Handles marine weather data integration from Open-Meteo API (free, no API key required)
"""

import logging
from datetime import datetime, timedelta
from typing import Dict, Optional

import aiohttp
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_async_session
from app.models import WeatherData

logger = logging.getLogger(__name__)


class WeatherService:
    """Service for fetching and caching marine weather data using Open-Meteo"""

    def __init__(self):
        # Open-Meteo API - completely free, no API key required
        self.base_url = "https://api.open-meteo.com/v1/forecast"
        self.cache_duration = timedelta(hours=1)  # Cache weather data for 1 hour

    async def get_weather_data(self, latitude: float, longitude: float, session: AsyncSession) -> Optional[Dict]:
        """
        Get weather data for a location, using cache if available

        Args:
            latitude: Location latitude
            longitude: Location longitude
            session: Database session

        Returns:
            Weather data dictionary or None if failed
        """
        try:
            # Check cache first
            cached_data = await self._get_cached_weather(latitude, longitude, session)
            if cached_data:
                return cached_data

            # Fetch from API
            weather_data = await self._fetch_weather_from_api(latitude, longitude)

            if weather_data:
                # Cache the data
                await self._cache_weather_data(latitude, longitude, weather_data, session)
                return weather_data

        except Exception as e:
            logger.error(f"Error getting weather data for {latitude}, {longitude}: {e}")

        return None

    async def _get_cached_weather(self, latitude: float, longitude: float, session: AsyncSession) -> Optional[Dict]:
        """Get cached weather data if still valid"""
        try:
            cutoff_time = datetime.utcnow() - self.cache_duration

            result = await session.execute(
                select(WeatherData)
                .where(
                    WeatherData.latitude == latitude,
                    WeatherData.longitude == longitude,
                    WeatherData.created_at >= cutoff_time
                )
                .order_by(WeatherData.created_at.desc())
                .limit(1)
            )

            weather_record = result.scalar_one_or_none()
            if weather_record:
                return weather_record.data

        except Exception as e:
            logger.error(f"Error checking weather cache: {e}")

        return None

    async def _fetch_weather_from_api(self, latitude: float, longitude: float) -> Optional[Dict]:
        """Fetch weather data from Open-Meteo API (free, no API key required)"""
        try:
            params = {
                'latitude': latitude,
                'longitude': longitude,
                'current': 'temperature_2m,relative_humidity_2m,apparent_temperature,weather_code,wind_speed_10m,wind_direction_10m,precipitation',
                'timezone': 'UTC',
                'wind_speed_unit': 'kmh',
                'temperature_unit': 'celsius'
            }

            async with aiohttp.ClientSession() as client_session:
                async with client_session.get(self.base_url, params=params) as response:
                    if response.status == 200:
                        data = await response.json()
                        current = data.get('current', {})

                        # Convert wind speed from km/h to knots for marine compatibility
                        wind_speed_kph = current.get('wind_speed_10m', 0)
                        wind_speed_knots = wind_speed_kph / 1.852

                        # Interpret WMO weather code
                        weather_code = current.get('weather_code', 0)
                        weather_desc = self._get_weather_description(weather_code)

                        # Estimate wave height from wind speed (simplified marine model)
                        wave_height = min(wind_speed_kph / 20, 4)
                        
                        weather_info = {
                            'temperature': current.get('temperature_2m'),
                            'humidity': current.get('relative_humidity_2m'),
                            'wind_speed': wind_speed_knots,  # in knots
                            'wind_speed_kph': wind_speed_kph,  # in km/h
                            'wind_direction': current.get('wind_direction_10m'),
                            'wind_direction_cardinal': self._degrees_to_cardinal(current.get('wind_direction_10m', 0)),
                            'visibility': 10,  # Default good visibility
                            'pressure': current.get('pressure'),
                            'weather_main': weather_desc,
                            'weather_description': weather_desc,
                            'wave_height': wave_height,
                            'sea_state': self._calculate_sea_state(wind_speed_kph, wave_height),
                            'precipitation': current.get('precipitation', 0),
                            'is_raining': self._is_raining_code(weather_code),
                            'marine_conditions': self._assess_marine_conditions(current),
                            'sts_optimality': self._calculate_sts_optimality(wind_speed_knots, wave_height, current.get('temperature_2m', 20)),
                            'last_updated': datetime.utcnow().isoformat()
                        }

                        return weather_info

        except Exception as e:
            logger.error(f"Error fetching weather from Open-Meteo API: {e}")

        return None

    def _assess_marine_conditions(self, current: Dict) -> str:
        """Assess marine conditions based on weather data"""
        try:
            wind_speed_kph = current.get('wind_speed_10m', 0)
            precipitation = current.get('precipitation', 0)
            weather_code = current.get('weather_code', 0)

            # Beaufort scale for wind (in km/h)
            if wind_speed_kph < 1:
                wind_condition = "calm"
            elif wind_speed_kph < 6:
                wind_condition = "light"
            elif wind_speed_kph < 12:
                wind_condition = "moderate"
            elif wind_speed_kph < 20:
                wind_condition = "fresh"
            elif wind_speed_kph < 31:
                wind_condition = "strong"
            else:
                wind_condition = "storm"

            # Precipitation assessment
            if self._is_raining_code(weather_code):
                precip_condition = "raining"
            else:
                precip_condition = "dry"

            # Overall marine conditions
            if wind_condition in ["storm", "strong"]:
                return "hazardous"
            elif wind_condition == "fresh" or precip_condition == "raining":
                return "challenging"
            elif wind_condition == "moderate":
                return "fair"
            else:
                return "favorable"

        except Exception as e:
            logger.error(f"Error assessing marine conditions: {e}")
            return "unknown"

    def _get_weather_description(self, code: int) -> str:
        """Convert WMO Weather code to description"""
        weather_codes = {
            0: "Clear sky",
            1: "Mainly clear",
            2: "Partly cloudy",
            3: "Overcast",
            45: "Foggy",
            48: "Depositing rime fog",
            51: "Light drizzle",
            53: "Moderate drizzle",
            55: "Dense drizzle",
            61: "Slight rain",
            63: "Moderate rain",
            65: "Heavy rain",
            71: "Slight snow",
            73: "Moderate snow",
            75: "Heavy snow",
            77: "Snow grains",
            80: "Slight rain showers",
            81: "Moderate rain showers",
            82: "Violent rain showers",
            85: "Slight snow showers",
            86: "Heavy snow showers",
            95: "Thunderstorm",
            96: "Thunderstorm with slight hail",
            99: "Thunderstorm with heavy hail",
        }
        return weather_codes.get(code, "Unknown")

    def _is_raining_code(self, code: int) -> bool:
        """Check if weather code indicates rain"""
        return code >= 51 and code <= 82

    def _calculate_sea_state(self, wind_kph: float, wave_height: float) -> str:
        """Calculate sea state from wind speed and wave height"""
        if wind_kph < 5 or wave_height < 0.5:
            return "calm"
        elif wind_kph < 15 or wave_height < 1.25:
            return "slight"
        elif wind_kph < 25 or wave_height < 2.5:
            return "moderate"
        elif wind_kph < 35 or wave_height < 4:
            return "rough"
        return "very_rough"

    def _degrees_to_cardinal(self, degrees: float) -> str:
        """Convert degrees to cardinal direction"""
        directions = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", 
                     "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
        index = int((degrees + 11.25) / 22.5) % 16
        return directions[index]

    def _calculate_sts_optimality(self, wind_knots: float, wave_height: float, temperature: float) -> Dict:
        """Calculate STS operation optimality score (0-100)"""
        score = 100

        # Wind Assessment (Critical)
        if wind_knots > 35:
            score = 0  # Stop operations
        elif wind_knots > 30:
            score -= 40  # Marginal conditions
        elif wind_knots > 20:
            score -= 20  # Monitor
        elif wind_knots > 15:
            score -= 10  # Good conditions degrading

        # Wave Height Assessment
        if wave_height > 4:
            score -= 40
        elif wave_height > 3:
            score -= 20
        elif wave_height > 2:
            score -= 10

        # Temperature extremes
        if temperature < 5 or temperature > 40:
            score -= 5

        return {
            "optimal": score >= 80,
            "percentage": max(0, score)
        }

    async def _cache_weather_data(self, latitude: float, longitude: float, data: Dict, session: AsyncSession):
        """Cache weather data in database"""
        try:
            weather_record = WeatherData(
                latitude=latitude,
                longitude=longitude,
                data=data
            )
            session.add(weather_record)
            await session.commit()

        except Exception as e:
            logger.error(f"Error caching weather data: {e}")

    async def get_weather_for_location(self, location_name: str, session: AsyncSession) -> Optional[Dict]:
        """
        Get weather for a named location (requires geocoding)

        Args:
            location_name: Location name (e.g., "Singapore Strait")
            session: Database session

        Returns:
            Weather data or None
        """
        # This would require a geocoding service to convert location names to lat/lon
        # For now, return None - would need to implement geocoding
        logger.warning(f"Location-based weather not implemented for: {location_name}")
        return None


# Global weather service instance
weather_service = WeatherService()

File: src/services/WeatherService.ts
/**
 * Weather Service
 * Handles fetching and caching weather data for STS operations
 * 
 * Uses weatherapi.com free API
 * Consider upgrade for production use
 */

export interface WeatherData {
  temperature: number; // Celsius
  temperatureF: number; // Fahrenheit
  feelsLike: number; // Celsius
  windSpeed: number; // knots
  windSpeedKph: number; // km/h
  windDirection: string; // Cardinal direction (N, NE, E, etc)
  windDirectionDegrees: number; // 0-360
  humidity: number; // percentage
  visibility: number; // km
  waveHeight?: number; // meters (approximated)
  seaState: string; // calm, slight, moderate, rough, very_rough
  conditions: string; // Weather condition text
  precipitation: number; // mm
  isRaining: boolean;
  optimalForSTS: boolean;
  optimalPercentage: number; // 0-100
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export class WeatherService {
  private static readonly API_BASE = 'https://api.weatherapi.com/v1';
  private static readonly API_KEY = import.meta.env.VITE_WEATHER_API_KEY;
  private static cache: Map<string, {data: WeatherData; timestamp: number}> = new Map();
  private static readonly CACHE_DURATION = 5 * 60 * 1000; // 5 minutes

  /**
   * Get weather for coordinates
   * Implements caching to avoid excessive API calls
   */
  static async getWeather(coords: Coordinates): Promise<WeatherData> {
    const cacheKey = `${coords.latitude},${coords.longitude}`;

    // Check cache
    const cached = this.cache.get(cacheKey);
    if (cached && Date.now() - cached.timestamp < this.CACHE_DURATION) {
      return cached.data;
    }

    try {
      // If no API key, return mock data
      if (!this.API_KEY) {
        console.warn('Weather API key not configured, using mock data');
        return this.getMockWeather();
      }

      const response = await fetch(
        `${this.API_BASE}/current.json?key=${this.API_KEY}&q=${coords.latitude},${coords.longitude}&aqi=no`
      );

      if (!response.ok) {
        console.error('Weather API error:', response.status);
        return this.getMockWeather();
      }

      const jsonData = await response.json();
      const weatherData = this.parseWeatherResponse(jsonData);

      // Cache the result
      this.cache.set(cacheKey, {
        data: weatherData,
        timestamp: Date.now()
      });

      return weatherData;
    } catch (error) {
      console.error('Weather fetch error:', error);
      return this.getMockWeather();
    }
  }

  /**
   * Parse WeatherAPI response to our format
   */
  private static parseWeatherResponse(data: any): WeatherData {
    const current = data.current;
    const location = data.location;

    // Convert wind speed from km/h to knots
    const windSpeedKnots = current.wind_kph / 1.852;

    // Estimate wave height from wind speed and fetch (simplified)
    const waveHeight = Math.min(current.wind_kph / 20, 4);
    const seaState = this.calculateSeaState(current.wind_kph, waveHeight);

    // Calculate STS optimality
    const optimality = this.calculateSTSOptimality({
      windSpeedKnots,
      visibility: current.vis_km,
      waveHeight,
      isRaining: current.will_it_rain === 1,
      temperature: current.temp_c
    });

    return {
      temperature: current.temp_c,
      temperatureF: current.temp_f,
      feelsLike: current.feelslike_c,
      windSpeed: windSpeedKnots,
      windSpeedKph: current.wind_kph,
      windDirection: current.wind_dir,
      windDirectionDegrees: current.wind_degree,
      humidity: current.humidity,
      visibility: current.vis_km,
      waveHeight: waveHeight,
      seaState: seaState,
      conditions: current.condition.text,
      precipitation: current.precip_mm,
      isRaining: current.will_it_rain === 1,
      optimalForSTS: optimality.optimal,
      optimalPercentage: optimality.percentage
    };
  }

  /**
   * Calculate sea state from wind speed and wave height
   */
  private static calculateSeaState(windKph: number, waveHeight: number): string {
    if (windKph < 5 || waveHeight < 0.5) return 'calm';
    if (windKph < 15 || waveHeight < 1.25) return 'slight';
    if (windKph < 25 || waveHeight < 2.5) return 'moderate';
    if (windKph < 35 || waveHeight < 4) return 'rough';
    return 'very_rough';
  }

  /**
   * Calculate STS operation optimality (0-100 score)
   * Based on maritime industry standards for STS operations
   */
  private static calculateSTSOptimality(conditions: {
    windSpeedKnots: number;
    visibility: number;
    waveHeight: number;
    isRaining: boolean;
    temperature: number;
  }): {optimal: boolean; percentage: number} {
    let score = 100;

    // Wind Assessment (Critical)
    // STS ops typically require <20 knots, marginal 20-30 knots
    if (conditions.windSpeedKnots > 35) {
      score -= 100; // Stop operations
    } else if (conditions.windSpeedKnots > 30) {
      score -= 40; // Marginal conditions
    } else if (conditions.windSpeedKnots > 20) {
      score -= 20; // Monitor
    } else if (conditions.windSpeedKnots > 15) {
      score -= 10; // Good conditions degrading
    }

    // Visibility Assessment
    // Minimum safe visibility ~5km for manifold operations
    if (conditions.visibility < 2) {
      score -= 50; // Hazardous
    } else if (conditions.visibility < 5) {
      score -= 30; // Marginal
    } else if (conditions.visibility < 10) {
      score -= 10; // Acceptable
    }

    // Wave Height Assessment
    // Typically max 3-4m for STS operations
    if (conditions.waveHeight > 4) {
      score -= 40;
    } else if (conditions.waveHeight > 3) {
      score -= 20;
    } else if (conditions.waveHeight > 2) {
      score -= 10;
    }

    // Precipitation
    if (conditions.isRaining) {
      score -= 15;
    }

    // Temperature extremes (operational comfort)
    if (conditions.temperature < 5 || conditions.temperature > 40) {
      score -= 5;
    }

    return {
      optimal: score >= 80,
      percentage: Math.max(0, Math.round(score))
    };
  }

  /**
   * Get mock weather data (for testing or API unavailability)
   */
  private static getMockWeather(): WeatherData {
    return {
      temperature: 28,
      temperatureF: 82,
      feelsLike: 30,
      windSpeed: 12,
      windSpeedKph: 22,
      windDirection: 'NE',
      windDirectionDegrees: 45,
      humidity: 65,
      visibility: 10,
      waveHeight: 0.8,
      seaState: 'slight',
      conditions: 'Partly Cloudy',
      precipitation: 0,
      isRaining: false,
      optimalForSTS: true,
      optimalPercentage: 85
    };
  }

  /**
   * Clear cache manually if needed
   */
  static clearCache(): void {
    this.cache.clear();
  }

  /**
   * Get weather for common port locations
   */
  static async getWeatherByPortName(portName: string): Promise<WeatherData> {
    const ports: Record<string, Coordinates> = {
      'fujairah': { latitude: 25.1242, longitude: 56.3569 },
      'dubai': { latitude: 25.2048, longitude: 55.2708 },
      'singapore': { latitude: 1.3521, longitude: 103.8198 },
      'rotterdam': { latitude: 51.9225, longitude: 4.4792 },
      'shanghai': { latitude: 31.2304, longitude: 121.4737 },
      'hamburg': { latitude: 53.5511, longitude: 9.9937 },
    };

    const coords = ports[portName.toLowerCase()];
    if (!coords) {
      throw new Error(`Port ${portName} not found in database`);
    }

    return this.getWeather(coords);
  }
}

export default WeatherService;